File: apps/web/test/followed-stocks.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { FollowedStocksStore, isStorableAssetId } from '../src/account/followed-stocks.js';

const SUBJECT = 'did:privy:follower';

function body(revision: number, assetIds: readonly string[]): string {
  return JSON.stringify({
    schemaVersion: 1, revision, assetIds,
    updatedAt: revision === 0 ? null : '2026-09-19T10:00:00.000Z',
  });
}

const json = (text: string, status = 200): Response =>
  new Response(text, {status, headers: {'content-type': 'application/json'}});

function store(handler: (request: Request) => Promise<Response> | Response): FollowedStocksStore {
  return new FollowedStocksStore({
    subject: SUBJECT,
    currentSubject: () => SUBJECT,
    accessToken: async () => 'header.payload.signature',
    apiBaseUrl: 'https://api.example',
    fetch: (async (input: RequestInfo | URL, init?: RequestInit) =>
      handler(new Request(input as never, init))) as typeof fetch,
  });
}

test('nothing is read until something asks', async () => {
  let calls = 0;
  const followed = store(() => { calls += 1; return json(body(0, [])); });
  assert.equal(calls, 0);
  assert.deepEqual(followed.getSnapshot().assetIds, []);
  assert.equal(followed.getSnapshot().status, 'idle');
  await followed.refresh();
  assert.equal(calls, 1);
  assert.equal(followed.getSnapshot().status, 'ready');
  followed.close();
});

test('keeping an asset writes the whole list at the revision it saw', async () => {
  const seen: Array<{method: string; body?: unknown}> = [];
  let revision = 1;
  let stored: string[] = ['apple'];
  const followed = store(async request => {
    if (request.method === 'GET') {
      seen.push({method: 'GET'});
      return json(body(revision, stored));
    }
    const sent = await request.json() as Record<string, unknown>;
    seen.push({method: 'PUT', body: sent});
    assert.equal(sent['baseRevision'], revision);
    stored = sent['assetIds'] as string[];
    revision += 1;
    return json(body(revision, stored));
  });
  await followed.refresh();
  await followed.follow('tesla');
  assert.deepEqual(followed.getSnapshot().assetIds, ['apple', 'tesla']);
  assert.equal(followed.getSnapshot().errorCode, null);
  assert.deepEqual(seen.map(entry => entry.method), ['GET', 'PUT']);
  followed.close();
});

test('keeping something already kept writes nothing', async () => {
  let writes = 0;
  const followed = store(request => {
    if (request.method !== 'GET') writes += 1;
    return json(body(2, ['apple']));
  });
  await followed.refresh();
  await followed.follow('apple');
  assert.equal(writes, 0);
  assert.equal(followed.isFollowing('apple'), true);
  followed.close();
});

test('a concurrent change is reloaded and this tap is not lost', async () => {
  let serverRevision = 1;
  let stored: string[] = ['apple'];
  let refusals = 0;
  const followed = store(async request => {
    if (request.method === 'GET') return json(body(serverRevision, stored));
    const sent = await request.json() as Record<string, unknown>;
    if (sent['baseRevision'] !== serverRevision) {
      refusals += 1;
      return json(JSON.stringify({
        error: {code: 'WATCHLIST_REVISION_CONFLICT', message: 'x', requestId: 'r'},
        currentSnapshot: JSON.parse(body(serverRevision, stored)),
      }), 409);
    }
    stored = sent['assetIds'] as string[];
    serverRevision += 1;
    return json(body(serverRevision, stored));
  });
  await followed.refresh();
  // Another client adds one and moves the revision on.
  serverRevision = 6;
  stored = ['apple', 'nvidia'];
  await followed.follow('tesla');
  assert.equal(refusals, 1, 'the stale write must be refused once');
  assert.equal(followed.getSnapshot().errorCode, null, 'the retry succeeded');
  assert.deepEqual(followed.getSnapshot().assetIds, ['apple', 'nvidia', 'tesla'],
    'the other change survives and this one is applied on top');
  followed.close();
});

test('a full list is reported rather than silently dropping the tap', async () => {
  const full = Array.from({length: 50}, (_, index) => `asset-${index}`);
  const followed = store(request => {
    assert.equal(request.method, 'GET', 'a full list never writes');
    return json(body(3, full));
  });
  await followed.refresh();
  assert.equal(followed.getSnapshot().full, true);
  await followed.follow('tesla');
  assert.equal(followed.getSnapshot().errorCode, 'FOLLOWING_LIST_FULL');
  followed.close();
});

test('unkeeping removes exactly one asset', async () => {
  let stored: string[] = ['apple', 'tesla'];
  const followed = store(async request => {
    if (request.method === 'GET') return json(body(4, stored));
    stored = (await request.json() as Record<string, unknown>)['assetIds'] as string[];
    return json(body(5, stored));
  });
  await followed.refresh();
  await followed.unfollow('apple');
  assert.deepEqual(followed.getSnapshot().assetIds, ['tesla']);
  followed.close();
});

test('identifiers the server would refuse never reach the network', async () => {
  let calls = 0;
  const followed = store(() => { calls += 1; return json(body(0, [])); });
  for (const bad of ['Apple', 'apple ', '-apple', 'apple--x', 'a'.repeat(101), '']) {
    await followed.follow(bad);
    assert.equal(followed.getSnapshot().errorCode, 'FOLLOWING_INVALID_INPUT', bad);
  }
  assert.equal(calls, 0);
  assert.equal(isStorableAssetId('3m-company'), true);
  assert.equal(isStorableAssetId('Apple'), false);
  followed.close();
});

test('a malformed stored list is refused rather than shown', async () => {
  const followed = store(() => json(JSON.stringify({
    schemaVersion: 1, revision: 1, assetIds: ['Apple'], updatedAt: '2026-09-19T10:00:00.000Z',
  })));
  await followed.refresh();
  assert.equal(followed.getSnapshot().status, 'error');
  assert.deepEqual(followed.getSnapshot().assetIds, []);
  followed.close();
});

File: apps/mobile/test/followed_stocks_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:trimmy/markets/followed_stocks.dart';
import 'package:trimmy/markets/followed_stocks_controller.dart';
import 'package:trimmy/practice_sync/http_transport.dart';

const _account = '9f000000-0000-4000-8000-000000000001';
final _base = Uri.parse('https://api.example');

http.Response _json(Object body, [int status = 200]) => http.Response(
  jsonEncode(body),
  status,
  headers: const {'content-type': 'application/json'},
);

Map<String, Object?> _snapshot({
  int revision = 1,
  List<String> assetIds = const ['apple'],
  String? updatedAt = '2026-09-19T10:00:00.000Z',
}) => <String, Object?>{
  'schemaVersion': 1,
  'revision': revision,
  'assetIds': assetIds,
  'updatedAt': revision == 0 ? null : updatedAt,
};

HttpFollowedStocksClient _client(
  http.Client transport, {
  String account = _account,
}) => HttpFollowedStocksClient(
  client: transport,
  baseUri: _base,
  accountId: account,
  accessToken: () async =>
      PracticeAccessToken(accountId: _account, token: 'a.b.c'),
);

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  group('the followed list client', () {
    test('reads a snapshot and an empty list', () async {
      final client = _client(
        MockClient((request) async {
          expect(request.url.path, '/v1/following');
          expect(request.method, 'GET');
          return _json(_snapshot(revision: 0, assetIds: const []));
        }),
      );
      final list = await client.read();
      expect(list.revision, 0);
      expect(list.assetIds, isEmpty);
      expect(list.updatedAt, isNull);
    });

    test('writes the whole list at the revision it saw', () async {
      String? sent;
      final client = _client(
        MockClient((request) async {
          sent = request.body;
          return _json(
            _snapshot(revision: 2, assetIds: const ['apple', 'tesla']),
          );
        }),
      );
      final saved = await client.write(
        baseRevision: 1,
        assetIds: const ['apple', 'tesla'],
        mutationId: '11111111-1111-4111-8111-111111111111',
      );
      final body = jsonDecode(sent!) as Map<String, dynamic>;
      expect(body['schemaVersion'], 1);
      expect(body['baseRevision'], 1);
      expect(body['assetIds'], ['apple', 'tesla']);
      expect(saved.revision, 2);
    });

    test(
      'refuses identifiers the server would refuse, before sending',
      () async {
        var sends = 0;
        final client = _client(
          MockClient((request) async {
            sends += 1;
            return _json(_snapshot());
          }),
        );
        for (final bad in [
          ['Apple'],
          ['apple '],
          ['-apple'],
          ['apple--x'],
          ['a' * 101],
          ['apple', 'apple'],
        ]) {
          expect(
            () => client.write(
              baseRevision: 0,
              assetIds: bad,
              mutationId: '11111111-1111-4111-8111-111111111111',
            ),
            throwsA(isA<FollowedStocksException>()),
            reason: bad.toString(),
          );
        }
        expect(sends, 0, reason: 'nothing unusable reaches the network');
      },
    );

    test(
      'refuses a non-HTTPS origin and a token for another account',
      () async {
        expect(
          () => HttpFollowedStocksClient(
            client: MockClient((_) async => _json(_snapshot())),
            baseUri: Uri.parse('http://api.example'),
            accountId: _account,
            accessToken: () async =>
                PracticeAccessToken(accountId: _account, token: 'a.b.c'),
          ),
          throwsA(isA<FollowedStocksException>()),
        );
        final mismatched = HttpFollowedStocksClient(
          client: MockClient((_) async => _json(_snapshot())),
          baseUri: _base,
          accountId: _account,
          accessToken: () async => PracticeAccessToken(
            accountId: '9f000000-0000-4000-8000-000000000002',
            token: 'a.b.c',
          ),
        );
        await expectLater(
          mismatched.read(),
          throwsA(
            isA<FollowedStocksException>().having(
              (error) => error.failure,
              'failure',
              FollowedStocksFailure.accountMismatch,
            ),
          ),
        );
      },
    );

    test('separates the failures that share a status', () async {
      final cases =
          <({int status, String code, FollowedStocksFailure failure})>[
            (
              status: 401,
              code: 'WATCHLIST_UNAUTHENTICATED',
              failure: FollowedStocksFailure.unauthenticated,
            ),
            (
              status: 409,
              code: 'WATCHLIST_REVISION_CONFLICT',
              failure: FollowedStocksFailure.revisionConflict,
            ),
            (
              status: 409,
              code: 'WATCHLIST_IDEMPOTENCY_CONFLICT',
              failure: FollowedStocksFailure.invalidRequest,
            ),
            (
              status: 503,
              code: 'WATCHLIST_UNAVAILABLE',
              failure: FollowedStocksFailure.notConfigured,
            ),
            (
              status: 503,
              code: 'WATCHLIST_REVISION_EXHAUSTED',
              failure: FollowedStocksFailure.limitReached,
            ),
          ];
      for (final row in cases) {
        final client = _client(
          MockClient(
            (_) async => _json({
              'error': {'code': row.code, 'message': 'x', 'requestId': 'r'},
            }, row.status),
          ),
        );
        await expectLater(
          client.read(),
          throwsA(
            isA<FollowedStocksException>().having(
              (error) => error.failure,
              'failure',
              row.failure,
            ),
          ),
          reason: row.code,
        );
      }
    });
  });

  group('the followed list controller', () {
    test('following adds one asset and keeps the rest', () async {
      var revision = 1;
      var stored = <String>['apple'];
      final controller = FollowedStocksController(
        client: _client(
          MockClient((request) async {
            if (request.method == 'GET') {
              return _json(_snapshot(revision: revision, assetIds: stored));
            }
            final body = jsonDecode(request.body) as Map<String, dynamic>;
            expect(body['baseRevision'], revision);
            stored = (body['assetIds'] as List).cast<String>();
            revision += 1;
            return _json(_snapshot(revision: revision, assetIds: stored));
          }),
        ),
      );
      addTearDown(controller.dispose);
      await controller.refresh();
      expect(controller.isFollowing('tesla'), isFalse);
      await controller.follow('tesla');
      expect(controller.assetIds, ['apple', 'tesla']);
      expect(controller.failure, isNull);
    });

    test('following something already followed writes nothing', () async {
      var writes = 0;
      final controller = FollowedStocksController(
        client: _client(
          MockClient((request) async {
            if (request.method != 'GET') writes += 1;
            return _json(_snapshot(assetIds: const ['apple']));
          }),
        ),
      );
      addTearDown(controller.dispose);
      await controller.refresh();
      await controller.follow('apple');
      expect(writes, 0);
      expect(controller.assetIds, ['apple']);
    });

    test('unfollowing removes exactly one asset', () async {
      var stored = <String>['apple', 'tesla'];
      final controller = FollowedStocksController(
        client: _client(
          MockClient((request) async {
            if (request.method == 'GET') {
              return _json(_snapshot(revision: 3, assetIds: stored));
            }
            stored = ((jsonDecode(request.body) as Map)['assetIds'] as List)
                .cast<String>();
            return _json(_snapshot(revision: 4, assetIds: stored));
          }),
        ),
      );
      addTearDown(controller.dispose);
      await controller.refresh();
      await controller.unfollow('apple');
      expect(controller.assetIds, ['tesla']);
    });

    test('a concurrent change is reloaded and the tap is not lost', () async {
      // The list moved to revision 5 elsewhere while this client held 1.
      var serverRevision = 1;
      var stored = <String>['apple'];
      var refusals = 0;
      final controller = FollowedStocksController(
        client: _client(
          MockClient((request) async {
            if (request.method == 'GET') {
              return _json(
                _snapshot(revision: serverRevision, assetIds: stored),
              );
            }
            final body = jsonDecode(request.body) as Map<String, dynamic>;
            if (body['baseRevision'] != serverRevision) {
              refusals += 1;
              return _json({
                'error': {
                  'code': 'WATCHLIST_REVISION_CONFLICT',
                  'message': 'x',
                  'requestId': 'r',
                },
                'currentSnapshot': _snapshot(
                  revision: serverRevision,
                  assetIds: stored,
                ),
              }, 409);
            }
            stored = (body['assetIds'] as List).cast<String>();
            serverRevision += 1;
            return _json(_snapshot(revision: serverRevision, assetIds: stored));
          }),
        ),
      );
      addTearDown(controller.dispose);
      await controller.refresh();
      // Someone else adds one and moves the revision on.
      serverRevision = 5;
      stored = ['apple', 'nvidia'];
      await controller.follow('tesla');
      expect(refusals, 1, reason: 'the stale write must be refused once');
      expect(controller.failure, isNull, reason: 'the retry succeeded');
      expect(
        controller.assetIds,
        ['apple', 'nvidia', 'tesla'],
        reason: 'the other change is kept and this one is applied on top',
      );
    });

    test(
      'a full list is reported rather than silently dropping the tap',
      () async {
        final full = List<String>.generate(50, (index) => 'asset-$index');
        final controller = FollowedStocksController(
          client: _client(
            MockClient((request) async {
              expect(request.method, 'GET', reason: 'a full list never writes');
              return _json(_snapshot(revision: 2, assetIds: full));
            }),
          ),
        );
        addTearDown(controller.dispose);
        await controller.refresh();
        expect(controller.isFull, isTrue);
        await controller.follow('tesla');
        expect(controller.failure, FollowedStocksFailure.limitReached);
      },
    );

    test('nothing is read until something asks', () async {
      var reads = 0;
      final controller = FollowedStocksController(
        client: _client(
          MockClient((_) async {
            reads += 1;
            return _json(_snapshot());
          }),
        ),
      );
      addTearDown(controller.dispose);
      expect(reads, 0);
      expect(controller.loaded, isFalse);
      await controller.refresh();
      expect(reads, 1);
    });
  });
}
